File: module-11-rl-agents/solutions/trajectories.ts
/**
 * Module 11: Trajectory collection
 * Complete rollouts (state, action, reward) from agent episodes.
 */
import Anthropic from "@anthropic-ai/sdk";
import { AgentEnvironment, type EpisodeSetup } from "./environment.js";
import { computeReward, type TrajectoryStep, type EpisodeReward, type RewardConfig } from "./rewards.js";
import { CODING_AGENT_PROMPT } from "../../module-5-coding-agent/solutions/prompt.js";
import type { Message } from "../../module-2-agent-loop/solutions/types.js";
import type { Grader, Transcript, TranscriptTurn } from "../../module-7-evals/solutions/types.js";

const client = new Anthropic();

export interface Trajectory {
  episodeId: string;
  prompt: string;
  steps: TrajectoryStep[];
  transcript: Transcript;
  reward: EpisodeReward;
  success: boolean;
}

/**
 * Collect a complete trajectory by running the agent in the environment.
 */
export async function collectTrajectory(
  env: AgentEnvironment,
  setup: EpisodeSetup,
  grader: Grader,
  rewardConfig?: RewardConfig
): Promise<Trajectory> {
  const state = await env.reset(setup);
  const steps: TrajectoryStep[] = [];
  const transcriptTurns: TranscriptTurn[] = [];
  const messages: Message[] = [{ role: "user", content: setup.prompt }];
  transcriptTurns.push({ role: "user", content: setup.prompt });

  const startTime = Date.now();
  let totalTokens = 0;

  for (let turn = 0; turn < state.maxSteps; turn++) {
    const response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 4096,
      system: CODING_AGENT_PROMPT + `\nWorkspace: ${state.workspaceDir}`,
      tools: state.registry.getDefinitions(),
      messages,
    });

    totalTokens += response.usage.input_tokens + response.usage.output_tokens;
    messages.push({ role: "assistant", content: response.content });

    if (response.stop_reason === "end_turn") {
      const text = response.content.find((b) => b.type === "text");
      transcriptTurns.push({
        role: "assistant",
        content: text?.type === "text" ? text.text : "",
      });
      break;
    }

    const toolCalls: TranscriptTurn["toolCalls"] = [];
    const toolResults: Anthropic.ToolResultBlockParam[] = [];

    for (const block of response.content) {
      if (block.type === "tool_use") {
        const input = block.input as Record<string, unknown>;
        const { result } = await env.step(block.name, input);
        const wasValid = !result.startsWith("Error:");

        steps.push({ toolName: block.name, input, result, wasValid });
        toolCalls.push({ name: block.name, input, result });
        toolResults.push({ type: "tool_result", tool_use_id: block.id, content: result });
      }
    }

    transcriptTurns.push({ role: "assistant", content: "", toolCalls });
    messages.push({ role: "user", content: toolResults });
  }

  const transcript: Transcript = {
    task: setup.prompt,
    turns: transcriptTurns,
    totalTokens,
    durationMs: Date.now() - startTime,
  };

  // Grade the episode
  const grade = await grader.grade(state.workspaceDir, transcript);
  const reward = computeReward(grade, steps, rewardConfig);

  await env.cleanup();

  return {
    episodeId: `ep-${Date.now()}`,
    prompt: setup.prompt,
    steps,
    transcript,
    reward,
    success: grade.passed,
  };
}

File: module-11-rl-agents/solutions/python/trajectories.py
"""Module 11: Trajectory collection."""
import os
import sys
import time

sys.path.insert(0, os.path.dirname(__file__))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-5-coding-agent", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-7-evals", "solutions", "python"))

import anthropic
from environment import AgentEnvironment
from rewards import compute_reward, RewardConfig
from prompt import CODING_AGENT_PROMPT
from eval_types import Transcript, TranscriptTurn

client = anthropic.Anthropic()


def collect_trajectory(env: AgentEnvironment, setup: dict, grader, reward_config: RewardConfig = None) -> dict:
    state = env.reset(setup)
    steps = []
    transcript_turns = [TranscriptTurn(role="user", content=setup["prompt"])]
    messages = [{"role": "user", "content": setup["prompt"]}]

    start_time = time.time()
    total_tokens = 0

    for turn in range(state["max_steps"]):
        response = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=4096,
            system=CODING_AGENT_PROMPT + f"\nWorkspace: {state['workspace_dir']}",
            tools=state["registry"].get_definitions(),
            messages=messages,
        )

        total_tokens += response.usage.input_tokens + response.usage.output_tokens
        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason == "end_turn":
            text = ""
            for block in response.content:
                if block.type == "text":
                    text = block.text
                    break
            transcript_turns.append(TranscriptTurn(role="assistant", content=text))
            break

        tool_calls = []
        tool_results = []
        for block in response.content:
            if block.type == "tool_use":
                inp = block.input if isinstance(block.input, dict) else {}
                step_result = env.step(block.name, inp)
                was_valid = not step_result["result"].startswith("Error:")
                steps.append({"tool_name": block.name, "input": inp, "result": step_result["result"], "was_valid": was_valid})
                tool_calls.append({"name": block.name, "input": inp, "result": step_result["result"]})
                tool_results.append({"type": "tool_result", "tool_use_id": block.id, "content": step_result["result"]})

        transcript_turns.append(TranscriptTurn(role="assistant", content="", tool_calls=tool_calls))
        messages.append({"role": "user", "content": tool_results})

    transcript = Transcript(
        task=setup["prompt"], turns=transcript_turns,
        total_tokens=total_tokens, duration_ms=int((time.time() - start_time) * 1000),
    )

    grade = grader.grade(state["workspace_dir"], transcript)
    reward = compute_reward(grade, steps, reward_config)

    env.cleanup()

    return {
        "episode_id": f"ep-{int(time.time() * 1000)}",
        "prompt": setup["prompt"],
        "steps": steps,
        "transcript": transcript,
        "reward": reward,
        "success": grade.passed,
    }
